panoply/datasource.py

```diff
@@ -7,9 +10,12 @@ class DataSource(events.Emitter):
     options = None
 
     # data source constructor
-    def __init__(self, source, options = {}):
-        super(DataSource, self).__init__()
-        
+    def __init__(self, source, options = {}, events = {}):
+        super(DataSource, self).__init__(events)
+
+        if 'destination' not in source:
```

Comment:
First, this is very cryptic.  If we want to have a default behavior on our base class but still let the user of our class the ability to change this default behavior we can expose a function in our API say, ```getDestination()``` that by default will return ```source['type']``` and ask the user override it if he wants a different behavior.

Second, What if I am implementing a data source that I don't want its destination to be the source type. but still, to respect the destination if it was defined in the UI by the user.
Now what we are doing is:
```python
if 'destination' not in source:
    source['destination'] = "something other than source type"
````
1) You will break all the existing data source that this is what they do.
2) Let's assume we fixed the previous point, how I will change the destination to be something other than the source type but still to respect the user defined destination?  I don't know if the value I have defined by the user or by the base class.